app/status.go

```diff
@@ -282,6 +299,111 @@ func (a *App) SetStatusAwayIfNeeded(userID string, manual bool) {
 	a.SaveAndBroadcastStatus(status)
 }
 
+func (a *App) SetStatusDoNotDisturbSchedule(userId, currentDayOfTheWeek, currentTime string) {
+	statusSchedule, err := a.GetStatusSchedule(userId)
+
+	if err != nil {
+		return
+	}
+	startTime := ""
+	endTime := ""
+
+	switch currentDayOfTheWeek {
+	case "Mon":
+		startTime = statusSchedule.MondayStart
+		endTime = statusSchedule.MondayEnd
+	case "Tue":
+		startTime = statusSchedule.TuesdayStart
+		endTime = statusSchedule.TuesdayEnd
+	case "Wed":
+		startTime = statusSchedule.WednesdayStart
+		endTime = statusSchedule.WednesdayEnd
+	case "Thu":
+		startTime = statusSchedule.ThursdayStart
+		endTime = statusSchedule.ThursdayEnd
+	case "Fri":
+		startTime = statusSchedule.FridayStart
+		endTime = statusSchedule.FridayEnd
+	case "Sat":
+		startTime = statusSchedule.SaturdayStart
+		endTime = statusSchedule.SaturdayEnd
+	case "Sun":
+		startTime = statusSchedule.SundayStart
+		endTime = statusSchedule.SundayEnd
+	}
+
+	if startTime == "" && endTime == "" {
+		return
+	}
+
+	newLayout := "15:04"
+	check, _ := time.Parse(newLayout, currentTime)
+	start, _ := time.Parse(newLayout, startTime)
+	end, _ := time.Parse(newLayout, endTime)
+
+	if a.InTimeSpan(start, end, check) {
+		a.SetStatusOnline(userId, true)
+	} else {
+		a.SetStatusDoNotDisturbScheduled(userId)
+	}
+}
+
+func (a *App) InTimeSpan(start, end, check time.Time) bool {
+	if check.After(end) {
+		return false
+	}
+	if end.After(start) {
+		return check.After(start)
+	}
+	return check.Before(start)
+}
+
+func (a *App) SetStatusSchedule(userId, mondayStart, mondayEnd, tuesdayStart, tuesdayEnd, wednesdayStart, wednesdayEnd, thursdayStart, thursdayEnd, fridayStart, fridayEnd, saturdayStart, saturdayEnd, sundayStart, sundayEnd string, mode int64) {
+	if !*a.Config().ServiceSettings.EnableUserStatuses {
+		return
+	}
+	statusSchedule, err := a.GetStatusSchedule(userId)
+
+	if err != nil {
+		statusSchedule = &model.StatusSchedule{UserId: userId}
+	}
+
+	statusSchedule.MondayStart = mondayStart
+	statusSchedule.MondayEnd = mondayEnd
+	statusSchedule.TuesdayStart = tuesdayStart
+	statusSchedule.TuesdayEnd = tuesdayEnd
+	statusSchedule.WednesdayStart = wednesdayStart
+	statusSchedule.WednesdayEnd = wednesdayEnd
+	statusSchedule.ThursdayStart = thursdayStart
+	statusSchedule.ThursdayEnd = thursdayEnd
+	statusSchedule.FridayStart = fridayStart
+	statusSchedule.FridayEnd = fridayEnd
+	statusSchedule.SaturdayStart = saturdayStart
+	statusSchedule.SaturdayEnd = saturdayEnd
+	statusSchedule.SundayStart = sundayStart
+	statusSchedule.SundayEnd = sundayEnd
+	statusSchedule.Mode = mode
+
+	a.SaveStatusSchedule(statusSchedule)
+}
+
+func (a *App) SetStatusDoNotDisturbScheduled(userID string) {
```

Comment:
This method should handle errors and return them and then all invokers should handle them.